File: src/04-smart-analyzer/web_scraper.py
"""
Web情報収集モジュール
"""
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse
import time
from typing import Dict, List, Optional

class WebScraper:
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        })
    
    def scrape_url(self, url: str) -> Dict:
        """URLからコンテンツを取得"""
        try:
            response = self.session.get(url, timeout=10)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, 'html.parser')
            
            # タイトル取得
            title = soup.find('title')
            title_text = title.get_text().strip() if title else "No Title"
            
            # 本文取得（基本的なクリーニング）
            for script in soup(["script", "style", "nav", "footer", "header"]):
                script.decompose()
            
            content = soup.get_text()
            content = ' '.join(content.split())  # 余分な空白を削除
            
            return {
                "success": True,
                "url": url,
                "title": title_text,
                "content": content[:5000],  # 最初の5000文字
                "content_length": len(content),
                "scraped_at": time.time()
            }
        
        except requests.exceptions.RequestException as e:
            return {
                "success": False,
                "url": url,
                "error": f"HTTP Error: {str(e)}",
                "scraped_at": time.time()
            }
        except Exception as e:
            return {
                "success": False,
                "url": url,
                "error": f"Parse Error: {str(e)}",
                "scraped_at": time.time()
            }
    
    def extract_links(self, url: str, base_url: str = None) -> List[str]:
        """ページ内のリンクを抽出"""
        try:
            response = self.session.get(url, timeout=10)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, 'html.parser')
            base_url = base_url or url
            
            links = []
            for link in soup.find_all('a', href=True):
                href = link['href']
                full_url = urljoin(base_url, href)
                if self._is_valid_url(full_url):
                    links.append(full_url)
            
            return list(set(links))  # 重複除去
        
        except Exception as e:
            return []
    
    def _is_valid_url(self, url: str) -> bool:
        """URLの妥当性チェック"""
        try:
            result = urlparse(url)
            return all([result.scheme, result.netloc])
        except:
            return False

# グローバルインスタンス
scraper = WebScraper() 

File: src/07-smart-analyzer-ts/web-scraper.ts
/**
 * Web情報収集モジュール - TypeScript版
 */
import axios, { AxiosResponse } from 'axios';
import * as cheerio from 'cheerio';
import { URL } from 'url';

export interface ScrapeResult {
  success: boolean;
  url: string;
  title?: string;
  content?: string;
  content_length?: number;
  scraped_at: number;
  error?: string;
}

export class WebScraper {
  private axiosInstance;

  constructor() {
    this.axiosInstance = axios.create({
      timeout: 10000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
      }
    });
  }

  async scrapeUrl(url: string): Promise<ScrapeResult> {
    try {
      const response: AxiosResponse = await this.axiosInstance.get(url);
      
      const $ = cheerio.load(response.data);
      
      // タイトル取得
      const title = $('title').text().trim() || "No Title";
      
      // 不要な要素を削除
      $('script, style, nav, footer, header, .advertisement, .ads').remove();
      
      // 本文取得
      let content = $('body').text();
      content = content.replace(/\s+/g, ' ').trim(); // 余分な空白を削除
      
      return {
        success: true,
        url: url,
        title: title,
        content: content.substring(0, 5000), // 最初の5000文字
        content_length: content.length,
        scraped_at: Date.now()
      };
    } catch (error) {
      if (axios.isAxiosError(error)) {
        return {
          success: false,
          url: url,
          error: `HTTP Error: ${error.message}`,
          scraped_at: Date.now()
        };
      } else {
        return {
          success: false,
          url: url,
          error: `Parse Error: ${error instanceof Error ? error.message : String(error)}`,
          scraped_at: Date.now()
        };
      }
    }
  }

  async extractLinks(url: string, baseUrl?: string): Promise<string[]> {
    try {
      const response: AxiosResponse = await this.axiosInstance.get(url);
      const $ = cheerio.load(response.data);
      
      const base = baseUrl || url;
      const links: string[] = [];
      
      $('a[href]').each((_, element) => {
        const href = $(element).attr('href');
        if (href) {
          try {
            const fullUrl = new URL(href, base).toString();
            if (this.isValidUrl(fullUrl)) {
              links.push(fullUrl);
            }
          } catch {
            // 無効なURLは無視
          }
        }
      });
      
      // 重複除去
      return [...new Set(links)];
    } catch (error) {
      return [];
    }
  }

  private isValidUrl(url: string): boolean {
    try {
      const parsedUrl = new URL(url);
      return parsedUrl.protocol === 'http:' || parsedUrl.protocol === 'https:';
    } catch {
      return false;
    }
  }
}

// エクスポート用のインスタンス
export const scraper = new WebScraper();